plugins/module_utils/foreman_helper.py

```diff
@@ -643,10 +644,241 @@ def exit_json(self, changed=False, **kwargs):
+            else:
+                self.fail_json(msg='Scope {0} not found for {1}'.format(str(self.entity_scope), self.entity_name))
+        elif not self.entity_scope:
+            entity = self._resolve_entity(self.entity_name, value, entity_opts=self.entity_opts)
+        else:
+            self.fail_json(msg='Invalid scope {0} for {1}'.format(str(self.entity_scope), self.entity_name))
+
+        return entity
+
+    def _resolve_sub_entities(self, entity, entities, entity_dict):
+        for entity_name, entity_opts in {key: value for key, value in entities.items() if key in entity_dict}.items():
+            if 'scope' in entity_opts:
+                if entity_opts['scope'] == self.entity_name and entity:
+                    scope_id = entity['id']
+                elif entity_opts['scope'] in entity_dict:
+                    scope_id = entity_dict[entity_opts['scope']]['id']
+
+                scope = {'{}_id'.format(entity_opts['scope']): scope_id}
+            else:
+                scope = None
+            if entity_opts['type'] == 'entity' and type(entity_dict[entity_name]) is str:
+                entity_dict[entity_name] = self._resolve_entity(entity_name, entity_dict[entity_name], entity_opts, params=scope)
+            elif (entity_opts['type'] == 'entity_list' and type(entity_dict[entity_name]) is list
+                  and len(entity_dict[entity_name]) > 0 and type(entity_dict[entity_name][0]) is str):
+                entity_dict[entity_name] = self._resolve_entity_list(entity_name, entity_dict[entity_name], entity_opts, params=scope)
+
+    @property
+    def _unscoped_sub_entities(self):
+        return {key: value for key, value in self.sub_entities.items() if 'scope' not in value}
+
+    @property
+    def _scoped_sub_entities(self):
+        return {key: value for key, value in self.sub_entities.items() if 'scope' in value}
+
+    def _resolve_entity(self, entity_name, value, entity_opts, params=None):
+        p = apypie.Inflector()
+        resource_path = entity_opts.get('resource_type', p.pluralize(entity_name))
+        if resource_path == 'operatingsystems':
+            return self.find_operatingsystem(value, params=params,
+                                             failsafe=entity_opts.get('failsafe', False),
+                                             thin=entity_opts.get('thin', True))
+        else:
+            return self.find_resource_by(resource_path,
+                                         entity_opts.get('search_by', 'name'), value,
+                                         search_operator=entity_opts.get('search_operator', '='),
+                                         failsafe=entity_opts.get('failsafe', False),
+                                         thin=entity_opts.get('thin', True),
+                                         params=params)
+
+    def _resolve_entity_list(self, entity_name, values, entity_opts, params=None):
```

Comment:
same comments as for `_resolve_entity` apply here too, as the code is very similar.

I'd *love* to de-duplicate this, but I don't see a *good* way to do so w/o risking readability.

Reply:
You are right ! Done. Tell me if it's less readable

Reply:
hah, this looks good!